### PR TITLE
nixos/plasma: use upstream xsession/wayland files

### PR DESCRIPTION
**Pull request by pasqui23**
###### Motivation for this change
An updated version of https://github.com/NixOS/nixpkgs/pull/56988
Wayland still does not work and `~/.local/share/sddm/wayland-session.log` is empty after a failed Wayland login,as `journalctl -u display-manager`.

I will fix the override after making Wayland work

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:204]: this looks like something from the rebase. please revert

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:277]: same as https://github.com/NixOS/nixpkgs/pull/100057/files#r502532070

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:301]: https://github.com/NixOS/nixpkgs/pull/100057/files#r502532070

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:302]: https://github.com/NixOS/nixpkgs/pull/100057/files#r502532070

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:306]: https://github.com/NixOS/nixpkgs/pull/100057/files#r502532070

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:308]: https://github.com/NixOS/nixpkgs/pull/100057/files#r502532070

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:322]: https://github.com/NixOS/nixpkgs/pull/100057/files#r502532070

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:359]: trailing whitespace

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:10]: this being qt514 is intentional. guessing rebase artifact

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (pasqui23) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix:188]: I will,after making it work on wayland

**Comment by pasqui23** [comment]
So the wayland session still quit on startup,but now it does not seem to leave any traces,whether in the journal nor in ~/.local/share/sddm/wayland-session.log `.

@worldofpeace any ideas why?

**Comment by hameerabbasi** [comment]
Maybe the description of #44139 will help?

**Comment by pasqui23** [comment]
The wayland entry is present in sddm,but does not work.
Still ,I need to check with `QT_DEBUG_PLUGINS=1;QML_IMPORT_TRACE=1;`

**Comment by pasqui23** [comment]
Also need to check for `wrapQtApps`

**Comment by pasqui23** [comment]
@hameerabbasi Checked with the suggested patch,still not working,got in `.local/share/sddm/wayland-session.log `;
```
dbus-daemon[16640]: [session uid=1001 pid=16640] Activating service name='org.freedesktop.systemd1' requested by ':1.0' (uid=1001 pid=16644 comm="/nix/store/90n7mvl7n0zl98wx0jjx2syfma34xhka-dbus-1" label="kernel")
dbus-daemon[16640]: [session uid=1001 pid=16640] Activated service 'org.freedesktop.systemd1' failed: Process org.freedesktop.systemd1 exited with status 1
dbus-update-activation-environment: warning: error sending to systemd: org.freedesktop.DBus.Error.Spawn.ChildExited: Process org.freedesktop.systemd1 exited with status 1
qt.qpa.plugin: Could not load the Qt platform plugin "wayland-org.kde.kwin.qpa" in "" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: wayland-org.kde.kwin.qpa, wayland-egl, wayland, wayland-xcomposite-egl, wayland-xcomposite-glx, eglfs, linuxfb, minimal, minimalegl, offscreen, vnc, xcb.

"/nix/store/bkdyka5fa5f707a6hvx8l6v2s37wvwvd-kwin-5.18.5-bin/bin/kwin_wayland" ("--xwayland", "--libinput", "--exit-with-session=/nix/store/2vqimab57z8f8cpbzkax9cwnb1rl997j-plasma-workspace-5.18.5/libexec/startplasma-waylandsession") exited with code 6
startplasmacompositor: Shutting down...
startplasmacompositor: Done.
```

**Comment by matklad** [comment]
@pasqui23 see https://github.com/NixOS/nixpkgs/issues/43757#issuecomment-618710712

Specifically, `env KWIN_FORCE_OWN_QPA=1 kwin_wayland --version` fixes this particular error for me.

**Comment by pasqui23** [comment]
@matklad adding said variable to `environment.sessionVariables` still does not fix it for me

**Comment by CertainLach** [comment]
Adding `KWIN_FORCE_OWN_QPA=1` and increasing limit (0004-start_kdeinit-environ-hard-limit.patch) to `1<<20` makes `startplasma_wayland` work for me

**Comment by CertainLach** [comment]
For anyone wondering: plasma on wayland is not quite stable on NixOS, screenshots can't be taken because of `X-KDE-DBUS-Restricted-Interfaces` attribute absence, and plasmashell has some bugs (Task manager shows no apps) because of same reason

**Comment by pasqui23** [comment]
And finally the wayland session works!

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:374]: surely this shouldn't be here

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (jakobrs) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:192]: I believe this shouldn't be here (this is handled by 99d511124620cad08e1bac2c0ca29ea3222d7f5d, see [here](https://github.com/NixOS/nixpkgs/commit/99d511124620cad08e1bac2c0ca29ea3222d7f5d#diff-03fbf9f534327a96ff2a6a1b6b3aa40a397d3dd9cac950f5af0ed500316b7512)).

**Comment by jakobrs** [comment]
I can confirm that this works when rebased onto nixos-unstable (see 0835411ea6ee199fb13fa0e68b60323430706459, I had to resolve a merge conflict).

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (worldofpeace) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix]: What? It seems we've pushed removing this and my pull request added this. Without this there the generated config for the plasma5 iso won't have this configuration anymore. Please re add @pasqui23 ?

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (ttuegel) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:190]: Since we are removing the custom session script, we should also remove `startplasma` above, but it calls several scripts which must still run at start-up.

**Review by ttuegel — CHANGES_REQUESTED** [review]

**Comment by worldofpeace** [comment]
Also want to mention that we don't have testing for the plasma5 wayland session... but this isn't a straightforward task so I'd be fine if it's not committed in this pull request. However we will need a release note **highlight**.

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (jakobrs) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix]: @worldofpeace This PR is based on a Nixpkgs commit from October, before you implemented `system.nixos-generate-config.desktopConfiguration` in November, and hasn't been rebased since. This part of your commit was `git add`ed in this  unrelated PR by mistake.

Comment (worldofpeace) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix]: how confusing. github :grin:

**Comment by pasqui23** [comment]
Also, a weird issue occurs when using the wayland session: sometimes the right click menu appears as a new window,complete with titlebar decoration.

**Comment by pasqui23** [comment]
For curiosity, is kwin's `src/plugins` directory built by nixpkgs?

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (pasqui23) [reply nixos/modules/services/x11/desktop-managers/plasma5.nix:190]: Now startplasma is called in services.xserver.displayManager.setupCommands

**Comment by colemickens** [comment]
I've applied this to my nixpkgs and am currently using the wayland session. Very exciting, thank you for working on this!

**Comment by worldofpeace** [comment]
I'm gonna rebase, write a release note, and then open an issue to get a wayland test.

**Comment by worldofpeace** [comment]
I've been trying this in a VM but it wasn't working for me... Need to investigate, but I'm wondering if anyone in the thread was testing in a VM. It's relevant because this module handles setup that, if regressed, won't be apparent right away without a clean state.

**Comment by pasqui23** [comment]
No,I haven't test it in a VM @worldofpeace .

**Comment by colemickens** [comment]
If I understand the question, it is possible (and indeed likely) that I had not previously run Plasma on this computer, so I wouldn't have had existing state to migrate or anything. I don't normally use Plasma so I wouldn't be able to test much more extensively, personally.

**Comment by pasqui23** [comment]
I have tested it on my pc with my personal plasma configuration.
It does not look like it has most problems besides the one I've mentioned.
As I've said the setup is now handled in `services.xserver.displayManager.setupCommands`, which only xinit and xpra do not respect.

**Comment by oxalica** [comment]
Tested on my PC. It almost works but still have some critical issues for daily use.
- **Task Manager/Taskbar failed to list windows opened**. It shows nothing but pinned buttons. I can only switch between windows by `Alt-Tab`. Other components/widgets work fine.
- **Input method doesn't work.** 
- `services.xserver.dpi` is ignored. Not sure if this is intended. Is there any replacement option for wayland?
- Some KDE settings are lost and reset to the default, including mouse speed, animation speed and all shortcuts.
- Screenshot (Spectacle) doesn't work. Printed `Error: Could not take a screenshot`. Here's the console log. Not sure if it's an issue of spectacle or plasma.
```
Error calling KWin DBus interface: "org.kde.kwin.Screenshot.Error.NoAuthorized" "The process is not authorized to take a screenshot"
QPixmap::scaled: Pixmap is a null pixmap
qt.qpa.wayland: Wayland does not support QWindow::requestActivate()
qt.qpa.wayland: Wayland does not support QWindow::requestActivate()
```

**Comment by samueldr** [comment]
I have what I think is a working solution for a couple of your issues I'm cleaning up, @oxalica. I'll ping this issue with the PR proposing the changes.

**Comment by pasqui23** [comment]
@oxalica try with pipewire enabled for the screenshot issue

**Comment by oxalica** [comment]
Finally I got things to work.

> Input method doesn't work.

It can be fixed by manually set `QT_IM_MODULE`. All `environment.variables` are not set in wayland session, since `/etc/profile` is not respected.
Isn't start script running in a login shell? Do we need to manually source it [here](https://github.com/NixOS/nixpkgs/pull/100057/files#diff-03fbf9f534327a96ff2a6a1b6b3aa40a397d3dd9cac950f5af0ed500316b7512L100)?

> services.xserver.dpi is ignored. Not sure if this is intended. Is there any replacement option for wayland?

Setting `forceFontDPIWayland=144` under `[General]` section in `~/.config/kcmfonts` fixes it. `startplasma-wayland` will read this value and set `QT_WAYLAND_FORCE_DPI` for all programs (See [code](https://github.com/KDE/plasma-workspace/blob/62842ac6fd5571394e38360b345a7ad3a1187273/startkde/startplasma-wayland.cpp#L34-L38)). 

@pasqui23 
> try with pipewire enabled for the screenshot issue

It's fixed by #116549 without `pipewire`.

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (oxalica) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:141]: `startplasma-x11` defined above should also be removed.
https://github.com/NixOS/nixpkgs/blob/c3e415f68d20424610469e402d41a8cc0b399565/nixos/modules/services/x11/desktop-managers/plasma5.nix#L16

**Review by oxalica — CHANGES_REQUESTED** [review]

**Comment by oxalica** [comment]
After #110536, the analog/digital clock widget got broken on (only) ayland. :cry: 
Note that time of analog clock is incorrect and fixed. The digital clock is below the analog clock and it shows nothing.
Screenshot_20210321_021850

**Comment by peterhoeg** [comment]
Guys, I'm spamming some of the old KDE wayland issues/PRs to ensure we get some testers. This comment is written from a plasma wayland session spawned from sddm after applying this PR (and the one-liner mentioned in a comment) which basically gets it going: https://github.com/NixOS/nixpkgs/pull/117102

**Comment by samueldr** [comment]
@peterhoeg which one-liner?

And you didn't hit anything not working without #116549?

**Comment by peterhoeg** [comment]
> @peterhoeg which one-liner?

I have updated the PR's first post with all the details, but basically you just need to point sddm at the directory with the wayland sessions:

``` nix
services.xserver.displayManager.sddm.settings.Wayland.SessionDir =
  "${pkgs.plasma5Packages.plasma-workspace}/share/wayland-sessions";
```

> And you didn't hit anything not working without #116549?

The pager doesn't show icons for the running applications but otherwise no, not
really. It's not possible for me to use as a daily driver though until I sort out the
custom keyboard mappings, but that's not Plasma's fault.

**Comment by peterhoeg** [comment]
@samueldr , reading your linked PR, I'm guessing this is why the pager/task manager do not show any icons.

**Comment by peterhoeg** [comment]
@samueldr - with your PR the task manager/pager both work! hidpi support is still horrible  so I can't really try it out until I'm on a desktop again.

**Comment by peterhoeg** [comment]
@pasqui23, when using systemd to manage the session (necessary patches are in master but haven't hit unstable yet), I can use the upstream sessions for both x11 and wayland without modification:

```nix
services.xserver.displayManager.sddm.settings.X11.SessionDir =
  "${pkgs.plasma5Packages.plasma-workspace}/share/xsessions";

services.xserver.displayManager.sddm.settings.Wayland.SessionDir =
  "${pkgs.plasma5Packages.plasma-workspace}/share/wayland-sessions";
```

This is admittedly only tried out with a user that already has everything set up, so I don't know if things will break without all the setup handled by our currect launcher script. That just needs to be tested.

**But** this got me thinking. Now that both GNOME and KDE use systemd, we could also start running these various scripts through systemd which of course also allows to share them in a cleaner way.

First we ensure that plasma uses systemd:

```nix
systemd.user.services = {
  plasma-run-with-systemd = {
    description = "Run via systemd";
    wantedBy = [ "basic.target" ];
    environment.XDG_CONFIG_HOME = "$HOME/.config";
    serviceConfig.Type = "oneshot";
    script = ''
      set -eEuo pipefail

      # use systemd to manage the session
      echo -e "[General]\nsystemdBoot=true" > $XDG_CONFIG_HOME/startkderc
    '';
  };
```

Then we run what is currently handled in our session wrapper script:

```nix
  plasma-early-setup = {
    description = "Early Plasma setup";
    wantedBy = [ "graphical-session-pre.target" ];
    environment = {
      XDG_CACHE_HOME = "$HOME/.cache";
      XDG_CONFIG_HOME = "$HOME/.config";
    };
    serviceConfig.Type = "oneshot";
    script = ''
      set -eEuo pipefail

      # The KDE icon cache is supposed to update itself automatically, but it uses
      # the timestamp on the icon theme directory as a trigger. This doesn't work
      # on NixOS because the timestamp never changes. As a workaround, delete the
      # icon cache at login and session activation.
      # See also: http://lists-archives.org/kde-devel/26175-what-when-will-icon-cache-refresh.html
      rm -fv $XDG_CACHE_HOME/icon-cache.kcache

      # xdg-desktop-settings generates this empty file but
      # it makes kbuildsyscoca5 fail silently. To fix this
      # remove that menu if it exists.
      rm -fv ''${XDG_CONFIG_HOME}/menus/applications-merged/xdg-desktop-menu-dummy.menu

      # Qt writes a weird ‘libraryPath’ line to
      # ~/.config/Trolltech.conf that causes the KDE plugin
      # paths of previous KDE invocations to be searched.
      # Obviously using mismatching KDE libraries is potentially
      # disastrous, so here we nuke references to the Nix store
      # in Trolltech.conf.  A better solution would be to stop
      # Qt from doing this wackiness in the first place.
      trolltech_conf="''${XDG_CONFIG_HOME}/Trolltech.conf"
      if [ -e "$trolltech_conf" ]; then
        ${pkgs.gnused}/bin/sed -i "$trolltech_conf" -e '/nix\\store\|nix\/store/ d'
      fi

      # Remove the kbuildsyscoca5 cache. It will be regenerated
      # immediately after. This is necessary for kbuildsyscoca5 to
      # recognize that software that has been removed.
      rm -fv $XDG_CACHE_HOME/ksycoca*

    ${pkgs.plasma5Packages.kservice}/bin/kbuildsycoca5
    '';
  };
};
```

Something like the trolltech.conf patching could also be split into a separate
unit as it applies to all environments in which people could possibly run qt
applications. So one way to deal with this would be to define a
`graphical-early-setup.service` for generic stuff and
`plasma-early-setup.service`, `gnome-early-setup.service`, etc.

**Comment by pasqui23** [comment]
@peterhoeg please add this issue to the wayland project

**Comment by jansol** [comment]
This could use a rebase

File: nixos/doc/manual/release-notes/rl-2105.xml
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2105.xml:42]: Where are the changes in the md file?

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (SuperSandro2000) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:182]: Please revert.

File: nixos/modules/services/x11/desktop-managers/plasma5.nix
Comment (SuperSandro2000) [inline nixos/modules/services/x11/desktop-managers/plasma5.nix:190]:
```suggestion
      services.xserver.displayManager.sessionPackages = [ pkgs.libsForQt5.plasma-workspace ];
```
We don't want to use aliases.

File: nixos/doc/manual/release-notes/rl-2105.xml
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2105.xml:101]: Please add in code the rename so that people with aliases enabled don't need to do anything.

**Comment by pasqui23** [comment]
@SuperSandro2000 rebased this branch onto master

File: nixos/doc/manual/release-notes/rl-2111.section.md
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2111.section.md:20]:
```suggestion
- KDE Plasma now finally works on Wayland.
```

**Comment by pasqui23** [comment]
Now that https://github.com/NixOS/nixpkgs/issues/118650 is fixed, let's merge this so that we can land Plasma Wayland by 21.11

**Comment by CertainLach** [comment]
While #118650 fixed with most bugs resolved, #139459 should also be merged before stabilizing plasma wayland session

This patch changes detected app ids, and in case if someone adds window rules before applying this patch - those rules will be broken after merge

**Comment by pasqui23** [comment]
@peterhoeg can we merge this?

**Comment by peterhoeg** [comment]
Yes please!

**Comment by pasqui23** [comment]
Ok I've done it @peterhoeg

**Review by oxalica — CHANGES_REQUESTED** [review]

**Review by oxalica — APPROVED** [review]
Also waiting for the merge.

**Comment by mweinelt** [comment]
This change wasn't converted from markdown to docbook and broke eval on staging-next. This was already visible on the CI changes on this pull request.

https://github.com/NixOS/nixpkgs/runs/3863386514

**Comment by cab404** [comment]
> This change wasn't converted from markdown to docbook and broke eval on staging-next. This was already visible on the CI changes on this pull request.
> 
> https://github.com/NixOS/nixpkgs/runs/3863386514

fixed in #141818

**Comment by mohe2015** [comment]
I believe this regressed the clock widget of plasma on Xorg. Can somebody verify please? I bisected to this particular PR. Will test a revert of it shortly.

**Comment by mohe2015** [comment]
@pasqui23 @peterhoeg @oxalica What do you think?

**Comment by SuperSandro2000** [comment]
klauncher does not have ``__NIXOS_SET_ENVIRONMENT_DONE`` set for me on wayland.

----

Well, that is not working for bash.

```
- The option definition `programs.bash.enable' in `/etc/nixos/configuration.nix' no longer has any effect; please remove it.
```

**Comment by oxalica** [comment]
@mohe2015 
Could you try running `kwriteconfig5 --file startkderc --group General --key systemdBoot true` and relogin? This enables "run with systemd mode" for plasma. Maybe it can fix environment variables.

**Pull request merged**
